# Deep Learning - Question 3

####  Normalize the data before feeding the data to the model and check how the normalization change your accuracy (code given below).
1. from sklearn.preprocessing import StandardScaler
2. sc = StandardScaler()

In [1]:
#Loading the Libraries

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers.core import Dense
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

In [2]:
#Importing the data
df = pd.read_csv('Breas Cancer.csv')

In [3]:
#Finding the overal info about the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

##### Observations:
1. There are 32 columns
2. Only one column is categorical (diagnosis)

In [4]:
#Printing the first five rows
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [5]:
#Finding the unique value or diagnosis column since it is only categorical value in data
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

In [6]:
#Finding any values in Unnamed: 32 column
df['Unnamed: 32'].unique()

array([nan])

#### As we could see there are no values in Unnamed: 32 columns, so we can skip this feature

In [7]:
df = df.drop(['Unnamed: 32'],axis = 1)

In [8]:
#Converting the categorical data to int
convert = {"diagnosis": {"M": 0, "B": 1}}
cancer_data = df.replace(convert)
cancer_data.head()

id  diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302          0        17.99         10.38          122.80     1001.0   
1    842517          0        20.57         17.77          132.90     1326.0   
2  84300903          0        19.69         21.25          130.00     1203.0   
3  84348301          0        11.42         20.38           77.58      386.1   
4  84358402          0        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [13]:
#Copying the remaining features other than diagnosis to x
x = cancer_data.iloc[:,2:31]

In [14]:
#Copying the diagnosis feature(Predictable feature)
y = cancer_data['diagnosis']

In [15]:
#splitting the data into training and testing categories
X_train, X_test, Y_train, Y_test = train_test_split(x, y,test_size=0.25, random_state=87)

In [16]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=29, activation='relu')) # hidden layer
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer

#Performing compilation
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

#Printing Summary and Accuracy
print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
14/14 [==============================] - 0s 2ms/step - loss: 64.2918 - acc: 0.6197
Epoch 2/100
14/14 [==============================] - 0s 2ms/step - loss: 14.4216 - acc: 0.5094
Epoch 3/100
14/14 [==============================] - 0s 2ms/step - loss: 8.2576 - acc: 0.4061
Epoch 4/100
14/14 [==============================] - 0s 2ms/step - loss: 3.1531 - acc: 0.6502
Epoch 5/100
14/14 [==============================] - 0s 2ms/step - loss: 2.3888 - acc: 0.5822
Epoch 6/100
14/14 [==============================] - 0s 2ms/step - loss: 1.2613 - acc: 0.7371
Epoch 7/100
14/14 [==============================] - 0s 2ms/step - loss: 0.8166 - acc: 0.8075
Epoch 8/100
14/14 [==============================] - 0s 2ms/step - loss: 0.6625 - acc: 0.8169
Epoch 9/100
14/14 [==============================] - 0s 2ms/step - loss: 0.5766 - acc: 0.8380
Epoch 10/100
14/14 [==============================] - 0s 2ms/step - loss: 0.4696 - acc: 0.8615
Epoch 11/100
14/14 [==============================] - 0s 

14/14 [==============================] - 0s 2ms/step - loss: 0.1423 - acc: 0.9437
Epoch 88/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1942 - acc: 0.9272
Epoch 89/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1863 - acc: 0.9249
Epoch 90/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1329 - acc: 0.9507
Epoch 91/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1240 - acc: 0.9531
Epoch 92/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1473 - acc: 0.9343
Epoch 93/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1375 - acc: 0.9460
Epoch 94/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1195 - acc: 0.9484
Epoch 95/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1260 - acc: 0.9484
Epoch 96/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1337 - acc: 0.9390
Epoch 97/100
14/14 [==============================] - 0s 2ms/st

#### Observations  :
1. Loss     = 25%
2. Accuracy = 93%

## Noramalizing the data using standard scalar

In [17]:
#Copying the remaining features other than diagnosis to x
x = cancer_data.iloc[:,2:31]

In [18]:
#Normalising the data using Standard scaler
SC = StandardScaler()
SC.fit(x) #Fitting the data
x_normalization = SC.transform(x)

In [19]:
#Copying the diagnosis feature(Predictable feature)
y = cancer_data['diagnosis']

In [20]:
#splitting the data into training and testing categories
X_train, X_test, Y_train, Y_test = train_test_split(x_normalization, y,test_size=0.25, random_state=87)

In [21]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=29, activation='relu')) # hidden layer
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer

#Performing compilation
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

#Printing Summary and Accuracy
print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
14/14 [==============================] - 0s 2ms/step - loss: 0.6265 - acc: 0.5305
Epoch 2/100
14/14 [==============================] - 0s 2ms/step - loss: 0.4970 - acc: 0.7958
Epoch 3/100
14/14 [==============================] - 0s 2ms/step - loss: 0.4020 - acc: 0.9249
Epoch 4/100
14/14 [==============================] - 0s 2ms/step - loss: 0.3305 - acc: 0.9554
Epoch 5/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2780 - acc: 0.9577
Epoch 6/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2372 - acc: 0.9577
Epoch 7/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2058 - acc: 0.9577
Epoch 8/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1824 - acc: 0.9601
Epoch 9/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1636 - acc: 0.9601
Epoch 10/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1491 - acc: 0.9601
Epoch 11/100
14/14 [==============================] - 0s 2m

14/14 [==============================] - 0s 3ms/step - loss: 0.0292 - acc: 0.9906
Epoch 87/100
14/14 [==============================] - 0s 3ms/step - loss: 0.0288 - acc: 0.9906
Epoch 88/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0285 - acc: 0.9906
Epoch 89/100
14/14 [==============================] - 0s 3ms/step - loss: 0.0282 - acc: 0.9906
Epoch 90/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0278 - acc: 0.9906
Epoch 91/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0275 - acc: 0.9906
Epoch 92/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0273 - acc: 0.9906
Epoch 93/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0270 - acc: 0.9906
Epoch 94/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0268 - acc: 0.9906
Epoch 95/100
14/14 [==============================] - 0s 2ms/step - loss: 0.0265 - acc: 0.9930
Epoch 96/100
14/14 [==============================] - 0s 2ms/st

#### Observations  after normalizing the data:
1. Loss     = 14%
2. Accuracy = 97%

## Conclusion :
<h4> Accuracy was increased and loss of data was decreased after Normalizing the data </h4>